### PR TITLE
[Merged by Bors] - Update to go 1.20

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Go 1.19 is now EOL and won't be updated any more. This PR updates our codebase to 1.20.

## Changes
- update all uses of go to 1.20: `go.mod`, `Dockerfile`s, documentation
- instead of top-level deprecated `math/rand` functions use `crypto/rand` or an instance of `*rand.Rand` instead
- use `errors.Join` instead of `github.com/hashicorp/go-multierror`
- fix `staticcheck` and `lint` complaints

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
I would like to also upgrade to 1.21 after this is merged, but will do it in 2 steps to keep the changes easier to review and test.

File: rand/rand.go [hunk 1]
```diff
@@ -7,11 +7,6 @@ import (
 	"time"
 )
 
-func init() {
-	// initialize global pseudo random generator
```
Comment (poszu) [inline rand/rand.go:11]: Could you explain why is it removed? Is `math.rand` seeded automatically now?

Btw, do we really need this package?

**Review by poszu — APPROVED** [review]

File: rand/rand.go (hunk 1 above)
Comment (fasmat) [reply rand/rand.go:11]: From https://golang.org/doc/go1.20#math/rand :

> The [math/rand](https://golang.org/pkg/math/rand/) package now automatically seeds the global random number generator (used by top-level functions like Float64 and Int) with a random value, and the top-level [Seed](https://golang.org/pkg/math/rand/#Seed) function has been deprecated. Programs that need a reproducible sequence of random numbers should prefer to allocate their own random source, using rand.New(rand.NewSource(seed)).
> ...
> The top-level [Read](https://golang.org/pkg/math/rand/#Read) function has been deprecated. In almost all cases, [crypto/rand.Read](https://golang.org/pkg/crypto/rand/#Read) is more appropriate.

Comment (fasmat) [reply rand/rand.go:11]: I removed our own `rand` package. It was only used in a handful of places (only tests) and could easily be replaced with `math/rand` and `crypto/rand`

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**